### PR TITLE
Intructors refactor

### PR DESCRIPTION
**Pull request by drouba**
Add course_offering_instructor table to databas
Add a section in Instructors page to link instructors to course offerings and display these pairings

### CODE REVIEW

File: data/init.sql [hunk 1]
```diff
@@ -152,4 +152,14 @@ CREATE TABLE IF NOT EXISTS course_sections(
...
+
+CREATE TABLE IF NOT EXISTS course_offerings_instructors(
+	offering_id INT(11) UNSIGNED NOT NULL,
+	user_id INT(11) UNSIGNED NOT NULL,
+	FOREIGN KEY (user_id) REFERENCES users(id),
+	FOREIGN KEY(offering_id) REFERENCES course_offerings(id) ON DELETE CASCADE,
+	created_at TIMESTAMP,
+	updated_at TIMESTAMP,
+	PRIMARY KEY (offering_id, user_id)
+);
```
Comment (jamestjw) [inline data/init.sql:165]: We could make the PK unique

File: modules/models/course_offering_instructor.php [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+class CourseOfferingInstructor extends Record {
+
+    static protected $table_name = "course_offerings_instructors";
+    
+
+    public $offering_id;
+    public $user_id;
+	public $created_at;
+	public $updated_at;
```
Comment (jamestjw) [inline modules/models/course_offering_instructor.php:11]: whats up here 😂  lets be consistent and use spaces

File: public/instructors_list.php [hunk 3]
```diff
@@ -85,6 +110,73 @@
...
+                <tr>
+                    <th>Course Offering</th>
+                    <th>Instructor</th>
+                    <th>Created At</th>
+                </tr>
+            </thead>
+            <tbody>
+        <?php foreach ($course_assignment as $row) { 
+            $inst = User::find_by(array('id' => $row->user_id));
+            $off = CourseOffering::find_by(array('id' => $row->offering_id));
```
Comment (jamestjw) [inline public/instructors_list.php:126]: You shouldnt be doing this! You should define an association between `CourseOfferingInstructor` and the `User` & `CourseOffering` classes

File: public/instructors_list.php [hunk 4]
```diff
@@ -85,6 +110,73 @@
...
+    </form>
+<?php } else if (!$instructors) { ?>
+    <blockquote> No instructors in database </blockquote>
+<?php } else { ?>
+    <blockquote> No course offering in database </blockquote>
+<?php } ?>
+
+<?php } else {?>
+    <?php include "templates/header.php"; ?>
+    <h2>You do not have the credentials to view this page.</h2>
```
Comment (jamestjw) [inline public/instructors_list.php:177]: Inclusion of the header file here looks odd, why is this required?

File: data/init.sql (hunk 1 above)
Comment (drouba) [reply data/init.sql:165]: It should be unique though, because its a combination of 2 unique keys no? I mena, we could add a seperate ID key, but I feel we would never use it and it would be redundant no?

File: public/instructors_list.php [hunk 5]
```diff
@@ -27,12 +41,25 @@
...
+    $offering_instructor->user_id = $_POST['instructor_selection'];
+    
+
+    try {
+        $offering_instructor->save();
+    } catch (PDOException $error) {
+        echo "<br>" . $error->getMessage();
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (drouba) [reply public/instructors_list.php:56]: good point, I'll work on that

File: public/instructors_list.php (hunk 3 above)
Comment (drouba) [reply public/instructors_list.php:126]: you mean like do a join between the classes?

File: public/instructors_list.php (hunk 4 above)
Comment (jamestjw) [reply public/instructors_list.php:177]: well the header.php is already included once above right?

File: public/instructors_list.php (hunk 3 above)
Comment (jamestjw) [reply public/instructors_list.php:126]: er.. not exactly, I meant that you could do 'CourseOfferingInstructor' belongs to 'User', same for `CourseOffering`

File: data/init.sql (hunk 1 above)
Comment (jamestjw) [reply data/init.sql:165]: actually forget what I said, I got confused with something else

File: public/instructors_list.php (hunk 4 above)
Comment (jamestjw) [reply public/instructors_list.php:177]: ok I mean my point is, why not include it just once at the top?

File: public/instructors_list.php [hunk 6]
```diff
@@ -85,6 +110,73 @@
...
+<h2>Assign Instructors to courses</h2>
+<?php if($course_offering && $instructors) { ?>
+    <form method="post">
+        Instructor: 
+        <select Name="instructor_selection" id="instructor_selection">
+            <option value="">----Select----</option>
+        <?php foreach($instructors as $row) { ?>
+            <option value="<?php echo $row->id; ?>">
+            <?php echo $row->last_name." ".$row->first_name; ?>
```
[the remark this answers is not in the document]
Comment (drouba) [reply public/instructors_list.php]: done

[the remark this answers is not in the document]
Comment (drouba) [reply public/instructors_list.php]: same as above

File: public/instructors_list.php [hunk 7]
```diff
@@ -85,6 +110,73 @@
...
+                    <th>Instructor</th>
+                    <th>Created At</th>
+                </tr>
+            </thead>
+            <tbody>
+        <?php foreach ($course_assignment as $row) { 
+            $inst = User::find_by(array('id' => $row->user_id));
+            $off = CourseOffering::find_by(array('id' => $row->offering_id));
+
+            $inst_name = $inst->first_name." ".$inst->last_name;
```
[the remark this answers is not in the document]
Comment (drouba) [reply public/instructors_list.php]: I tried that, but it still gave me an error :(

File: public/instructors_list.php [hunk 8]
```diff
@@ -85,6 +111,72 @@
...
+                    <th>Created At</th>
+                </tr>
+            </thead>
+            <tbody>
+        <?php foreach ($course_assignment as $row) { 
+            //$inst = User::find_by(array('id' => $row->user_id));
+            //$off = CourseOffering::find_by(array('id' => $row->offering_id));
+
+            //$inst_name = $inst.get_full_name();
+            //$off_name = $off->course_offering_name;
```
Comment (jamestjw) [inline public/instructors_list.php:130]: dont forget to remove the comments

File: public/instructors_list.php [hunk 9]
```diff
@@ -1,15 +1,30 @@
...
+
+try {
+    $course_offering = CourseOffering::getAll();
 } catch (PDOException $error) {
-    echo $sql . "<br>" . $error->getMessage();
+    echo "<br>" . $error->getMessage();
 }
+
+try {
+    $course_assignment = CourseOfferingInstructor::getAll();
```
Comment (jamestjw) [inline public/instructors_list.php:23]: Ok so you need to also preload the `course_offering` and the `user` associations here

File: public/instructors_list.php [hunk 10]
```diff
@@ -85,6 +111,72 @@
...
+        <?php foreach ($course_assignment as $row) { 
+            //$inst = User::find_by(array('id' => $row->user_id));
+            //$off = CourseOffering::find_by(array('id' => $row->offering_id));
+
+            //$inst_name = $inst.get_full_name();
+            //$off_name = $off->course_offering_name;
+            ?>
+            <tr>
+                <td><?php echo $row->course_offering->course_offering_name; ?></td>
+                <td><?php echo $row->user->first_name." ".$row->user->last_name; ?></td>
```
Comment (jamestjw) [inline public/instructors_list.php:134]: Heres the reason behind my comment on preloading before

Comment (jamestjw) [reply public/instructors_list.php:134]: for instance, if we have this 
image

we dont want 4 differnt queries to get all this, we want it to be done in 2 queries, one for the users one for the courseofferings, imagine if we have 20 course offerings on this page :) that would imply that we 'd 40 queries here and it'd be a nightmare 
image

Comment (jamestjw) [reply public/instructors_list.php:134]: you dont have the logging capability in your branch yet you need to rebase from main

File: public/instructors_list.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (drouba) [reply public/instructors_list.php]: done

File: dev_cga.log [hunk 11]
```diff
@@ -0,0 +1,518 @@
+[March 24, 2022, 1:31 pm] SELECT * FROM loggedin WHERE user_digest = ? LIMIT 1 [['user_digest', '09911812cc482990e40979559073e11c']]
+[March 24, 2022, 1:31 pm] SELECT * FROM users WHERE is_instructor = ? [['is_instructor', '1']]
+[March 24, 2022, 1:31 pm] SELECT * FROM course_offerings
+[March 24, 2022, 1:31 pm] SELECT * FROM course_offerings_instructors
+[March 24, 2022, 1:31 pm] SELECT * FROM course_offerings WHERE id = ? LIMIT 1 [['id', '1']]
+[March 24, 2022, 1:31 pm] SELECT * FROM users WHERE id = ? LIMIT 1 [['id', '2']]
+[March 24, 2022, 1:32 pm] SELECT * FROM loggedin WHERE user_digest = ? LIMIT 1 [['user_digest', '09911812cc482990e40979559073e11c']]
```
Comment (jamestjw) [inline dev_cga.log:7]: hey no don't check in this file 😂

Comment (jamestjw) [reply dev_cga.log:7]: add it to the gitignore

**Review by jamestjw — APPROVED** [review]
ok remove the log file before merging this

File: dev_cga.log (hunk 11 above)
Comment (drouba) [reply dev_cga.log:7]: ok
